### PR TITLE
Mutation Strategy of Replacing Random Activation

### PR DESCRIPTION
**Pull request by soumyadipghosh**
- Adds MutationStrategy as an argument to RandomPairwiseExchange in LTFB for Neural Architecture Search
- Adds the mutation strategy of replacing random activation layers in the winning model after a tournament in LTFB

### CODE REVIEW

File: include/lbann/execution_algorithms/ltfb/mutation_strategy.hpp [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+public:
+  NullMutation() = default; 
+  void mutate(model&, const int&) override {}
+};
+
+// Replace activation layers
+class ReplaceActivation : public Cloneable<ReplaceActivation, MutationStrategy>
+{
+private:
+  std::vector<std::string> m_activation_types = {"relu", "tanh", "softmax", "elu", "leaky relu", "log softmax"};
```
Comment (timmoon10) [inline include/lbann/execution_algorithms/ltfb/mutation_strategy.hpp:68]: Softmax and log-softmax are questionable to use as activations, except at the end of the network.

File: include/lbann/execution_algorithms/ltfb/mutation_strategy.hpp [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+       }
+    }
+
+    // Generate two random numbers - one for new layer type and one for old layer name
+    int type_index = fast_rand_int(get_fast_generator(), m_activation_types.size());
+    int name_index = fast_rand_int(get_fast_generator(), activation_layer_names.size());
+
+    // Name of new layer
+    std::string new_layer_name = "new_" + m_activation_types[type_index] 
+                                        + std::to_string(step) + std::to_string(name_index);
```
Comment (timmoon10) [inline include/lbann/execution_algorithms/ltfb/mutation_strategy.hpp:136]: If you are replacing the layer, it should be fine to reuse the old name. Right now we'll get layer names like `new_new_new_layer123`.

File: include/lbann/execution_algorithms/ltfb/mutation_strategy.hpp [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+ 
+    // Replace layer
+    m.replace_layer(make_new_activation_layer(comm, m_activation_types[type_index], new_layer_name), 
+                    activation_layer_names[name_index]);
+
+    // Erase old layer from activation layer list
+    activation_layer_names.erase(activation_layer_names.cbegin() + name_index);
+
+    // Add new layer to activation layer list
+    activation_layer_names.push_back(new_layer_name);
```
Comment (timmoon10) [inline include/lbann/execution_algorithms/ltfb/mutation_strategy.hpp:149]: It isn't necessary to update this if we aren't caching it:
```suggestion
```

File: include/lbann/execution_algorithms/ltfb/random_pairwise_exchange.hpp [hunk 4]
```diff
@@ -194,6 +200,10 @@ class RandomPairwiseExchange final
    */
   std::unique_ptr<ExchangeStrategy> m_comm_algo;
 
+  /** @brief The strategy for mutation of a model - insert more details later
```
Comment (timmoon10) [inline include/lbann/execution_algorithms/ltfb/random_pairwise_exchange.hpp:203]:
```suggestion
  /** @brief The strategy for mutation of a model
```

File: src/execution_algorithms/ltfb/random_pairwise_exchange.cpp [hunk 5]
```diff
@@ -307,15 +312,15 @@ void RandomPairwiseExchange::select_next(model& m,
     // FIXME (trb 03/18/21): This is ... not great. We need to
     // unravel the "fake" polymorphism in the model non-hierarchy
     // soon.
-    using DAGModel = directed_acyclic_graph_model;
-    auto& local_model = dynamic_cast<DAGModel&>(m);
-    auto& partner_dag_model = dynamic_cast<DAGModel&>(*partner_model);
-    local_model = std::move(partner_dag_model);
+
```
Comment (timmoon10) [inline src/execution_algorithms/ltfb/random_pairwise_exchange.cpp:315]: We are currently always keeping the local model:
```suggestion
    using DAGModel = directed_acyclic_graph_model;
    auto& local_model = dynamic_cast<DAGModel&>(m);
    auto& partner_dag_model = dynamic_cast<DAGModel&>(*partner_model);
    local_model = std::move(partner_dag_model);
    
```

**Review by timmoon10 — CHANGES_REQUESTED** [review]
Overall looks good. I do see we always choose the local model in LTFB, I presume for debugging. We should revert that change.

File: include/lbann/execution_algorithms/ltfb/mutation_strategy.hpp [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+    std::vector<std::string> activation_layer_names;
+
+    auto& comm = *m.get_comm();
+    
+    for (int i = 0; i < m.get_num_layers(); ++i)
+    {
+       // Creating a temp string with lower case representation
+       std::string temp_type = m.get_layer(i).get_type();
+       std::transform(temp_type.begin(), temp_type.end(),
+                                         temp_type.begin(), ::tolower);
```
Comment (benson31) [inline include/lbann/execution_algorithms/ltfb/mutation_strategy.hpp:122]:
```suggestion
       std::transform(begin(temp_type),
                      end(temp_type),
                      begin(temp_type),
                      [](unsigned char c){ return ::tolower(c); });    
```

File: include/lbann/execution_algorithms/ltfb/mutation_strategy.hpp [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+            lbann::elu_layer<DataType, data_layout::DATA_PARALLEL, Dev>>(&comm, 1);
+    } else if (new_type == "leaky relu") {
+       layer = std::make_unique<
+            lbann::leaky_relu_layer<DataType, data_layout::DATA_PARALLEL, Dev>>(&comm, 0.01);
+    } else if (new_type == "log softmax") {
+       layer = std::make_unique<
+            lbann::log_softmax_layer<DataType, data_layout::DATA_PARALLEL, Dev>>(&comm);
+    } else {
+       LBANN_ERROR("Unknown new layer type: ", new_type);
+    }
```
Comment (benson31) [inline include/lbann/execution_algorithms/ltfb/mutation_strategy.hpp:106]: Shouldn't this inherit the type, layout, and device allocation of the layer it's replacing?

File: include/lbann/execution_algorithms/ltfb/mutation_strategy.hpp [hunk 8]
```diff
@@ -0,0 +1,156 @@
...
+
+       if (std::find(m_activation_types.cbegin(), m_activation_types.cend(), temp_type)
+                                                              != m_activation_types.cend()) {
+         activation_layer_names.push_back(m.get_layer(i).get_name());
+       }
+    }
+
+    // Generate two random numbers - one for new layer type and one for old layer name
+    int type_index = fast_rand_int(get_fast_generator(), m_activation_types.size());
+    int name_index = fast_rand_int(get_fast_generator(), activation_layer_names.size());
```
Comment (benson31) [inline include/lbann/execution_algorithms/ltfb/mutation_strategy.hpp:132]:
```suggestion
    int const type_index = fast_rand_int(get_fast_generator(), m_activation_types.size());
    int const name_index = fast_rand_int(get_fast_generator(), activation_layer_names.size());
```

File: src/execution_algorithms/ltfb/random_pairwise_exchange.cpp (hunk 5 above)
Comment (benson31) [reply src/execution_algorithms/ltfb/random_pairwise_exchange.cpp:315]: This isn't right. Restoring this means that `local_model`, which is an alias for `m`, will now be set to the `partner_model` (i.e., the winning model in this case). But the documentation says that it's the *losing* model that gets mutated.

In summary: the way that it's written currently, without committing @timmoon10's suggestion, the losing model, which is the local model, gets mutated and returns. This breaks the current LTFB implementation in a fairly obvious way. Merging @timmoon10's suggestion restores the normal LTFB behavior when using the `NullMutation` strategy, but breaks the documented behavior when using `ReplaceActivation`. There's clearly a deeper logic error that needs to be resolved here.

**Review by benson31 — CHANGES_REQUESTED** [review]
1. The `mutation_strategy.hpp` header should not contain implementations. There are no templates, so please move everything to a suitably located `.cpp` file.
2. Run `clang-format` on everything.
3. There are some severe algorithmic issues that need to be addressed beyond the simple C++ issues.

File: include/lbann/execution_algorithms/ltfb/mutation_strategy.hpp (hunk 1 above)
Comment (soumyadipghosh) [reply include/lbann/execution_algorithms/ltfb/mutation_strategy.hpp]: Removed softmax, log-softmax; Added sigmoid

File: include/lbann/execution_algorithms/ltfb/mutation_strategy.hpp (hunk 2 above)
Comment (soumyadipghosh) [reply include/lbann/execution_algorithms/ltfb/mutation_strategy.hpp]: Keeping layer name same.

File: include/lbann/execution_algorithms/ltfb/mutation_strategy.hpp (hunk 3 above)
Comment (soumyadipghosh) [reply include/lbann/execution_algorithms/ltfb/mutation_strategy.hpp]: Removed

File: src/execution_algorithms/ltfb/random_pairwise_exchange.cpp (hunk 5 above)
Comment (soumyadipghosh) [reply src/execution_algorithms/ltfb/random_pairwise_exchange.cpp]: Yes, fixed this issue. The logic now is that if a model loses in a tournament, the winning model is copied over to it and then mutated.

File: python/lbann/core/training_algorithm.py [hunk 9]
```diff
@@ -186,6 +186,29 @@ def do_export_proto(self):
         params.local_training_algorithm.CopyFrom(self.local_algo.export_proto())
         return params
 
+class MutationStrategy:
+    """The strategy for mutation after a tournament in LTFB.
+       INSERT FURTHER DESCRIPTION HERE.
```
[the remark this answers is not in the document]
Comment (soumyadipghosh) [reply python/lbann/core/training_algorithm.py]: Added description.

File: include/lbann/execution_algorithms/ltfb/mutation_strategy.hpp (hunk 6 above)
Comment (soumyadipghosh) [reply include/lbann/execution_algorithms/ltfb/mutation_strategy.hpp]: done.

File: include/lbann/execution_algorithms/ltfb/mutation_strategy.hpp (hunk 7 above)
Comment (soumyadipghosh) [reply include/lbann/execution_algorithms/ltfb/mutation_strategy.hpp]: Has to be resolved at compile time, so cant use getter functions.

File: src/execution_algorithms/ltfb/random_pairwise_exchange.cpp [hunk 10]
```diff
@@ -134,23 +136,26 @@ bool local_wins(EvalType local,
...
 
 RandomPairwiseExchange::RandomPairwiseExchange(
   std::string metric_name,
   metric_strategy winner_strategy,
-  std::unique_ptr<ExchangeStrategy> comm_algo)
+  std::unique_ptr<ExchangeStrategy> comm_algo,
+  std::unique_ptr<MutationStrategy> mutate_algo)
   : RandomPairwiseExchange({{metric_name, winner_strategy}},
-                           std::move(comm_algo))
+                           std::move(comm_algo), std::move(mutate_algo))
```
Comment (benson31) [inline src/execution_algorithms/ltfb/random_pairwise_exchange.cpp:152]:
```suggestion
                           std::move(comm_algo),
                           std::move(mutate_algo))
```

File: src/execution_algorithms/ltfb/random_pairwise_exchange.cpp [hunk 11]
```diff
@@ -134,23 +136,26 @@ bool local_wins(EvalType local,
...
 {}
 
 RandomPairwiseExchange::RandomPairwiseExchange(
   RandomPairwiseExchange const& other)
-  : m_metrics{other.m_metrics}, m_comm_algo{other.m_comm_algo->clone()}
+  : m_metrics{other.m_metrics}, m_comm_algo{other.m_comm_algo->clone()},
```
Comment (benson31) [inline src/execution_algorithms/ltfb/random_pairwise_exchange.cpp:157]:
```suggestion
  : m_metrics{other.m_metrics},
    m_comm_algo{other.m_comm_algo->clone()},
```

File: include/lbann/execution_algorithms/ltfb/mutation_strategy.hpp [hunk 12]
```diff
@@ -0,0 +1,79 @@
...
+   *  @param[in] new_type The type of the new activation layer (ReLU or tanh
+   * etc)
+   *  @param[in] new_name The name of the new activation layer
+   */
+
+  std::unique_ptr<lbann::Layer>
+  make_new_activation_layer(lbann::lbann_comm& comm,
+                            std::string const& new_type,
+                            std::string const& new_name);
+  void mutate(model& m, const int& step);
```
Comment (benson31) [inline include/lbann/execution_algorithms/ltfb/mutation_strategy.hpp:74]:
```suggestion
  void mutate(model& m, const int& step) final;
```

File: include/lbann/execution_algorithms/ltfb/mutation_strategy.hpp [hunk 13]
```diff
@@ -0,0 +1,79 @@
...
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+// implied. See the License for the specific language governing
+// permissions and limitations under the license.
+////////////////////////////////////////////////////////////////////////////////
+#ifndef LBANN_EXECUTION_ALGORITHMS_LTFB_MUTATION_STRATEGY_HPP_INCLUDED
+#define LBANN_EXECUTION_ALGORITHMS_LTFB_MUTATION_STRATEGY_HPP_INCLUDED
+
+#include "lbann/layers/layer.hpp"
```
Comment (benson31) [inline include/lbann/execution_algorithms/ltfb/mutation_strategy.hpp:29]:
```suggestion
#include "lbann/models/model.hpp"
#include "lbann/utils/cloneable.hpp"
```
Once `make_new_activation_layer` is moved out of this file, it won't (explicitly) depend on `std::string`, `lbann::lbann_comm` or `lbann::Layer` anymore. This is the correct include set; however, the `model.hpp` could be replaced with a forward declaration.

**Review by benson31 — CHANGES_REQUESTED** [review]

**Review by benson31 — APPROVED** [review]

**Review by timmoon10 — APPROVED** [review]
If this passes Bamboo, then it's good to go.

**Review by timmoon10 — CHANGES_REQUESTED** [review]
We've identified a bug where different procs in a trainer can choose to replace different layers.

**Comment by soumyadipghosh** [comment]
Addressed the bug mentioned above. Pinging @timmoon10

**Review by timmoon10 — APPROVED** [review]
Looks good. If [the Bamboo build](https://lc.llnl.gov/bamboo/browse/LBANN-TIM392-2) has no new failures, it's ready to go.

**Pull request merged**